circuit/program/src/data/literal/cast.rs

```diff
@@ -0,0 +1,287 @@
+            Self::I128(integer) => cast_integer_to_type(integer, to_type),
+            Self::U8(integer) => cast_integer_to_type(integer, to_type),
+            Self::U16(integer) => cast_integer_to_type(integer, to_type),
+            Self::U32(integer) => cast_integer_to_type(integer, to_type),
+            Self::U64(integer) => cast_integer_to_type(integer, to_type),
+            Self::U128(integer) => cast_integer_to_type(integer, to_type),
+            Self::Scalar(scalar) => cast_scalar_to_type(scalar, to_type),
+            Self::Signature(..) => bail!("Cannot cast a signature literal to another type."),
+            Self::String(..) => bail!("Cannot cast a string literal to another type."),
+        }
+    }
+
+    /// Casts the literal to the given literal type, with lossy truncation.
+    ///
+    /// This method makes a *best-effort* attempt to preserve all bits of information,
+    /// but it is not guaranteed to do so.
+    ///
+    /// The hierarchy of casting is as follows:
+    ///  - (`Address`, `Group`) <-> `Field` <-> `Scalar` <-> `Integer` <-> `Boolean`
+    ///  - `String` (not supported)
+    /// Note that casting to left along the hierarchy always preserves information.
+    pub fn cast_lossy(&self, to_type: LiteralType) -> Result<Self> {
+        match self {
+            Self::Address(address) => cast_lossy_group_to_type(&address.to_group(), to_type),
+            Self::Boolean(boolean) => cast_lossy_boolean_to_type(boolean, to_type),
+            Self::Field(field) => cast_lossy_field_to_type(field, to_type),
+            Self::Group(group) => cast_lossy_group_to_type(group, to_type),
+            Self::I8(integer) => cast_lossy_integer_to_type(integer, to_type),
+            Self::I16(integer) => cast_lossy_integer_to_type(integer, to_type),
+            Self::I32(integer) => cast_lossy_integer_to_type(integer, to_type),
+            Self::I64(integer) => cast_lossy_integer_to_type(integer, to_type),
+            Self::I128(integer) => cast_lossy_integer_to_type(integer, to_type),
+            Self::U8(integer) => cast_lossy_integer_to_type(integer, to_type),
+            Self::U16(integer) => cast_lossy_integer_to_type(integer, to_type),
+            Self::U32(integer) => cast_lossy_integer_to_type(integer, to_type),
+            Self::U64(integer) => cast_lossy_integer_to_type(integer, to_type),
+            Self::U128(integer) => cast_lossy_integer_to_type(integer, to_type),
+            Self::Scalar(scalar) => cast_lossy_scalar_to_type(scalar, to_type),
+            Self::Signature(..) => bail!("Cannot cast a signature literal to another type."),
+            Self::String(..) => bail!("Cannot cast a string literal to another type."),
+        }
+    }
+}
+
+/// Casts a boolean literal to the given literal type.
+fn cast_boolean_to_type<A: Aleo>(boolean: &Boolean<A>, to_type: LiteralType) -> Result<Literal<A>> {
+    match to_type {
+        LiteralType::Address => Ok(Literal::Address(boolean.cast())),
```

Comment:
These `cast` and cast_lossy` calls seem easily macro-able. Might be better for cleanliness